File: Week_1/credit.c
#include <stdio.h>
#include <cs50.h>

bool visa_test(long num);
bool amex_test(long num);
bool mc_test(long num);
bool luns_test(long card_num);

int main(void)
{
    long number = get_long("Number: ");
    
    if(luns_test(number)){
       if(mc_test(number))
        {
            printf("MASTERCARD\n");
        }
        else if(visa_test(number))
        {
            printf("VISA\n");
        }
        else if(amex_test(number))
        {
            printf("AMEX\n");
        }
        else
        {
            // can pass but not be ne of these three card companies
            printf("INVALID\n");    
        }
    }
    else
    {
        printf("INVALID\n");    
    }
}

bool visa_test(long num)
{
    // visas can be 13 or 16 numbers and always start with 4, if after dividing by 1e13 or 1e16 and removing float portions i get 4, then it is a valid number.
    int testOne = num / 1000000000000;
    int testTwo = num / 1000000000000000;
    
    if(testOne == 4 || testTwo == 4)
    {
        return true;
    }
    
    return false;
}

bool amex_test(long num)
{   // if remainder after dividing bu 1e15 is 34 or 37 it is valid
    int test = num / 10000000000000;
    
    if(test == 34 || test == 37)
    {
        return true;
    }
    
    return false;
}

bool mc_test(long num)
{   // if after dividing bu 1e16 remainder is 51, 52, 53, 54, or 55 test passes
    int test = num / 100000000000000;
    
    if(test == 51 || test == 52 || test == 53 || test == 54 || test == 55)
    {
        return true;
    }
    
    return false;
}

bool luns_test(long card_num)
{
    // takes in a card number
    int odd_sum = 0, even_sum = 0, digit, digit_hold, test, last;
    do
    {
        // pull out last digit, any 2+ digit number will leave last by %
        digit = card_num % 10;
        // all odd index digits from left to right get addded together
        odd_sum += digit;
        
        // remove last digit
        card_num = (card_num - digit) / 10;

        // pull new to last digit
        digit = card_num % 10;
        // all even index digits are doubled then all of their digits are added together.
        digit_hold = digit * 2;

        // if digit became 2 after doubling, add per the same opperation as above.
        if(digit_hold > 9)
        {
            last = digit_hold % 10;
            even_sum += last;
            last = (digit_hold - last) / 10;
            even_sum += last;
        }
        else
        {
            // if digit did not 'overflow' add to even digits
            even_sum += digit_hold;
        }
        
        // remove last digit
        card_num = (card_num - digit) / 10;
    } // loop until the card number 'disapears'
    while(card_num > 0);

    // add evens and odds together
    test = odd_sum + even_sum;

    // if sum is multiple of 10 it is a valid number
    if(test % 10 == 0)
    {
        return true;
    }
    return false;
}

File: Week_6/credit.py
from cs50 import get_int

def main():
    while True:
        num = get_int("Number: ")
        if num > 0:
            break

    if luhns_test(num):
        s = str(num)
        l = len(s)
        if is_amex(s, l):
            print("AMEX")
        if is_mastercard(s, l):
            print("MASTERCARD")
        if is_visa(s, l):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")

def luhns_test(num):
    even = 0
    odd = 0
    while num > 0:
        even += num % 10
        num = strip_digit(num)

        if num == 0:
            break

        digit = (num % 10) * 2
        num = strip_digit(num)

        while digit > 0:
            odd += digit % 10
            digit = strip_digit(digit)

    if (even + odd) % 10 == 0:
        return True
    return False

def strip_digit(num):
    return (num - (num % 10)) / 10

def is_amex(s, l):
    if l == 15:
        if s[:2] == "37" or s[:2] == "34":
            return True
    return False

def is_mastercard(s, l):
    if l == 16:
        if int(s[:2]) >= 51 and int(s[:2]) <= 55:
            return True
    return False

def is_visa(s, l):
    if l == 16 or l == 13:
        if s[:1] == "4":
            return True
    return False

main()
